service/common/src/main/java/org/apache/polaris/service/catalog/iceberg/CatalogHandlerUtils.java

```diff
@@ -433,39 +450,203 @@ protected TableMetadata commit(TableOperations ops, UpdateTableRequest request)
+                      .requirements()
+                      .forEach((requirement) -> requirement.validate(baseWithRemovedSnaps));
+                } catch (CommitFailedException e) {
+                  throw new ValidationFailureException(e);
                 }
 
-                // commit
+                TableMetadata.Builder newMetadataBuilder = TableMetadata.buildFrom(newBase);
+                request.updates().forEach((update) -> update.applyTo(newMetadataBuilder));
+                TableMetadata updated = newMetadataBuilder.build();
+                // always commit this
                 taskOps.commit(base, updated);
               });
-
     } catch (ValidationFailureException e) {
       throw e.wrapped();
     }
 
     return ops.current();
   }
 
+  private UpdateRequirement.AssertRefSnapshotID findAssertRefSnapshotID(
+      UpdateTableRequest request) {
+    UpdateRequirement.AssertRefSnapshotID assertRefSnapshotID = null;
+    int total = 0;
+    for (UpdateRequirement requirement : request.requirements()) {
+      if (requirement instanceof UpdateRequirement.AssertRefSnapshotID) {
+        ++total;
+        assertRefSnapshotID = (UpdateRequirement.AssertRefSnapshotID) requirement;
+      }
+    }
+
+    // if > 1 assertion for refs, then it's not safe to roll back, make this Noop.
+    return total != 1 ? null : assertRefSnapshotID;
+  }
+
+  private List<MetadataUpdate> generateUpdatesToRemoveNoopSnapshot(
+      TableMetadata base,
+      long expectedCurrentSnapshotId,
+      String updateRefName,
+      long newSnapshotSeqNumber) {
+    // find the all the snapshots we want to retain which are not the part of current branch.
+    Set<Long> idsToRetain = Sets.newHashSet();
+    for (Map.Entry<String, SnapshotRef> ref : base.refs().entrySet()) {
+      String refName = ref.getKey();
+      SnapshotRef snapshotRef = ref.getValue();
+      if (refName.equals(updateRefName)) {
+        continue;
+      }
+      idsToRetain.add(ref.getValue().snapshotId());
+      // Always check the ancestry for both branch and tags
+      // mostly for case where a branch was created and then was dropped
+      // then a tag was created and then rollback happened post that tag
+      // was dropped and branch was re-created on it.
+      for (Snapshot ancestor : SnapshotUtil.ancestorsOf(snapshotRef.snapshotId(), base::snapshot)) {
+        idsToRetain.add(ancestor.snapshotId());
+      }
+    }
+
+    List<MetadataUpdate> updateToRemoveSnapshot = new ArrayList<>();
+    Long snapshotId = base.ref(updateRefName).snapshotId(); // current tip of the given branch
+    // ensure this branch has the latest sequence number.
+    long expectedSequenceNumber = base.lastSequenceNumber();
+    Set<Long> snapshotsToRemove = new LinkedHashSet<>();
+    while (snapshotId != null && !Objects.equals(snapshotId, expectedCurrentSnapshotId)) {
+      Snapshot snap = base.snapshot(snapshotId);
+      if (expectedSequenceNumber != snap.sequenceNumber()) {
```

Comment:
When would this occur?

Reply:
This can happen with multiple branch scenarios, where the commit seq number are not conitnous, added a comment.

Reply:
Might be worth adding a debug log